### PR TITLE
refactor: extract Configure and Metadata framework functions into single implementation

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Small refactor that removes boilerplate definition of `Metadata` and `Configure` functions in all our plugin-framework resources and data sources into common structs.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

### CODE REVIEW

File: mongodbatlas/fw_common.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+func (d *DSCommon) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	client, err := ConfigureClient(req.ProviderData)
+	if err != nil {
+		resp.Diagnostics.AddError(errorConfigureSummary, err.Error())
+		return
+	}
+	d.client = client
+}
+
+func ConfigureClient(providerData any) (*MongoDBClient, error) {
```
Comment (marcosuma) [inline mongodbatlas/fw_common.go]: Not sure if I am wrong here, but would you consider making this private if it's only being called in this package?

File: mongodbatlas/fw_common.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	if providerData == nil {
+		return nil, nil
+	}
+
+	client, ok := providerData.(*MongoDBClient)
+	if !ok {
+		return nil, fmt.Errorf(errorConfigure, providerData)
+	}
+
+	return client, nil
```
Comment (marcosuma) [inline mongodbatlas/fw_common.go]: nit:
```
if client, ok := providerData.(*MongoDBClient); ok {
    return client, nil
}

return nil, fmt.Errorf(...)
```

File: mongodbatlas/fw_common.go (hunk 1 above)
Comment (marcosuma) [inline mongodbatlas/fw_common.go]: any chance we could avoid using `any`?

File: mongodbatlas/fw_common.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+// common struct for all framework resources. Implements the following plugin-framework defined functions:
+// - Metadata
+// - Configure
+// client is left empty and populated by the framework when envoking Configure method.
+// resourceName must be defined when creating an instance of a resource.
```
Comment (andreaangiolillo) [inline mongodbatlas/fw_common.go]: comments should start with the name of the Struct or function

`// RSCommon is used ...`

File: mongodbatlas/fw_common.go (hunk 1 above)
Comment (andreaangiolillo) [reply mongodbatlas/fw_common.go]: +1, ConfigureClient should be called only here

File: mongodbatlas/fw_common.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package mongodbatlas
```
Comment (andreaangiolillo) [inline mongodbatlas/fw_common.go:1]: Does it make sense to move this struct in a subdir to make it clear that does not define a resource or ds?

**Review by andreaangiolillo — APPROVED** [review]
LGTM. My comments are not blocking the PR

File: mongodbatlas/fw_common.go (hunk 1 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_common.go]: for sure

File: mongodbatlas/fw_common.go (hunk 2 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_common.go]: will adjust

File: mongodbatlas/fw_common.go (hunk 1 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_common.go]: Not in this case. The ConfigureRequest struct defined by terraform plugin framework defines the `ProviderData` field as `any` becuase the data is provider-specifc.

File: mongodbatlas/fw_common.go (hunk 4 above)
Comment (AgustinBettati) [reply mongodbatlas/fw_common.go:1]: Had looked into this but reached a limitation. When defining `fw_common.go` within framework package, I need to import mongodbatlas package to reference `MongoDBClient`. This leads to an `import cycle not allowed` error.

**Review by maastha — APPROVED** [review]
LGTM, thank you for this one!

**Pull request merged**